events/mbed_events.cpp

```diff
@@ -0,0 +1,75 @@
+/* events
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "events/mbed_events.h"
+#include "mbed.h"
+
+using namespace events;
+
+namespace mbed {
+
+#ifdef MBED_CONF_RTOS_PRESENT
+/* Create an event queue, and start the thread that dispatches it. Static
+ * variables mean this happens once the first time each template instantiation
+ * is called. This is currently instantiated no more than twice.
+ */
+template
+<osPriority Priority, size_t QueueSize, size_t StackSize>
+EventQueue *do_shared_event_queue_with_thread()
+{
+    static unsigned char queue_buffer[QueueSize];
+    static EventQueue queue(sizeof queue_buffer, queue_buffer);
+
+    static unsigned char stack[StackSize];
+    static Thread thread(Priority, StackSize, stack);
+
+    Thread::State state = thread.get_state();
+    if (state == Thread::Inactive || state == Thread::Deleted) {
+        osStatus status = thread.start(callback(&queue, &EventQueue::dispatch_forever));
+
+        if (status != osOK) {
+            error("thread start status %d", status);
+            MBED_ASSERT(status == osOK);
+            return NULL;
+        }
+    }
+
+    return &queue;
+}
+#endif
+
+EventQueue *get_shared_normal_event_queue()
+{
+#if MBED_CONF_EVENTS_SHARED_NORMAL_DISPATCH_FROM_APPLICATION || !defined MBED_CONF_RTOS_PRESENT
+    /* Only create the EventQueue, but no dispatching thread */
+    static unsigned char queue_buffer[MBED_CONF_EVENTS_SHARED_NORMAL_EVENTSIZE];
+    static EventQueue queue(sizeof queue_buffer, queue_buffer);
+
+    return &queue;
+#else
+    return do_shared_event_queue_with_thread<osPriorityNormal, MBED_CONF_EVENTS_SHARED_NORMAL_EVENTSIZE, MBED_CONF_EVENTS_SHARED_NORMAL_STACKSIZE>();
+#endif
+}
+
+#ifdef MBED_CONF_RTOS_PRESENT
+EventQueue *get_shared_interrupt_deferral_event_queue()
+{
+    return do_shared_event_queue_with_thread<osPriorityRealtime, MBED_CONF_EVENTS_SHARED_IRQDEFERRAL_EVENTSIZE, MBED_CONF_EVENTS_SHARED_IRQDEFERRAL_STACKSIZE>();
```

Comment:
Not necessarily. Assuming you have 3 styles of drivers:
1. Drivers which require real-time latency and have short running events
2. Drivers which don't require real-time and have short running  events
3. Drivers which don't require real-time and have long running events

The current implementation allows an  mbed 2 application to use **type 2 and 3 drivers**. Trying to use a type 1 driver will cause a compiler error.

If you have both normal and interrupt queue present then  applications will be able to use **type 2 and 3 drivers** or **type 1 and 2 drivers**, but not type 1 and 3 at the same time. The biggest downside to this approach is you get a runtime assert error for mixing type 1 and 3 drivers, rather than getting an upfront compiler error.

I'm happy with keeping the interrupt queue out of mbed 2, as this is more conservative and doesn't risk setting any false expectations. I just wanted to make sure all the options explored.